[stdlib/msvc] Runtime with MSVC library

tinysun212 opened the pull request:
<!-- Please complete this template before creating pull request. -->
#### What's in this pull request?

<!-- Description about pull request. -->
#### Resolved bug number: ([SR-](https://bugs.swift.org/browse/SR-))

<!-- If this pull request resolves any bugs from Swift bug tracker -->

---

<!-- This selection should only be completed by Swift admin -->

Before merging this pull request to apple/swift repository:
- [ ] Test pull request on Swift continuous integration.

<details>
  <summary>

Triggering Swift CI</summary>



The swift-ci is triggered by writing a comment on this PR addressed to the GitHub user @swift-ci. Different tests will run depending on the specific comment that you use. The currently available comments are:

**Smoke Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please smoke test |
| OS X platform | @swift-ci Please smoke test OS X platform |
| Linux platform | @swift-ci Please smoke test Linux platform |

 **Validation Testing**

| Platform | Comment |
| --- | --- |
| All supported platforms | @swift-ci Please test |
| OS X platform | @swift-ci Please test OS X platform |
| Linux platform | @swift-ci Please test Linux platform |

Note: Only members of the Apple organization can trigger swift-ci.
</details>

<!-- Thank you for your contribution to Swift! -->

This patch is for libswiftCore.lib, linking with the library set of Visual Studio 2015. Clang with the option -fms-extension is used to build this port.

tinysun212 committed:
stdlib/msvc: Runtime with MSVC library

This patch is for libswiftCore.lib, linking with the library set of Visual Studio 2015. Clang with the option -fms-extension is used to build.

tkremenek commented:
@swift-ci smoke test

tkremenek commented:
Patch LGTM.

@gribozavr Can you review?

gribozavr commented on stdlib/public/stubs/Assert.cpp:
```diff
@@ -89,9 +121,20 @@ _swift_stdlib_reportUnimplementedInitializer(const char *className,
   char *log;
+#if defined(_MSC_VER)
+  int len =
+      _scprintf("fatal error: use of unimplemented "
```
Please define a shim for `asprintf()` instead of duplicating the code.

gribozavr commented on stdlib/public/stubs/LibcShims.cpp:
```diff
@@ -21,14 +26,25 @@
 #endif
 
+#if defined(_MSC_VER)
+static_assert(std::is_same<long long, swift::__swift_ssize_t>::value,
```
This assertion is correct as is.  Please fix the definition of `__swift_ssize_t` instead.

gribozavr commented on stdlib/public/stubs/LibcShims.cpp:
```diff
@@ -55,10 +71,14 @@ int _swift_stdlib_close(int fd) { return close(fd); }
+#elif defined(__GNU_LIBRARY__) || defined(__CYGWIN__) || defined(_MSC_VER)
 #include <malloc.h>
 size_t _swift_stdlib_malloc_size(const void *ptr) {
+#if defined(_MSC_VER)
```
Please do another `#elif` on the top level in the existing chain instead of splitting the function definition with `#if`.

gribozavr commented on stdlib/public/stubs/Stubs.cpp:
```diff
@@ -20,16 +20,25 @@
+#include <string>
+#include <iostream>
+#include <windows.h>
+// Macro max(), min() conflicts with std::max(), std::min()
```
Please `#define NOMINMAX` instead (per https://support.microsoft.com/en-us/kb/143208) -- here and in other files.

tinysun212 replied:
Thanks, I didn't know it.

gribozavr commented on stdlib/public/stubs/Stubs.cpp:
```diff
@@ -232,8 +241,19 @@ extern "C" uint64_t swift_float80ToString(char *Buffer, size_t BufferLength,
+  }
+  std::string Line;
+  std::getline(std::cin, Line);
+  *LinePtr = reinterpret_cast<char *>(std::malloc(Line.size() + 1));
```
Did you mean to call `memcpy()` somewhere?

Also, you shouldn't need to `reinterpret_cast`.

tinysun212 replied:
Without casting, ```assigning to 'char *' from incompatible type 'void *'``` displayed.
Did you mean to use ```static_cast<char *>``` or simply ```(char *)```?

gribozavr replied:
Sorry for being unclear, yes, `static_cast` would be better.

gribozavr commented on stdlib/public/stubs/LibcShims.cpp:
```diff
@@ -69,11 +89,27 @@ size_t _swift_stdlib_malloc_size(const void *ptr) {
+  return RandomUniform(MersenneRandom);
+#else
   return arc4random_uniform(upper_bound);
+#endif
```
Please consider submitting arc4random-related patches as a separate PR and without `#if`s, directly switching to portable C++.

tinysun212 replied:
Do you mean a PR of the code like this ?

```
#include <random>

static std::random_device RandomeDevice;
static std::mt19937 MersenneRandom(RandomeDevice());

__swift_uint32_t _swift_stdlib_arc4random(void) {
  return MersenneRandom();
}

__swift_uint32_t
_swift_stdlib_arc4random_uniform(__swift_uint32_t upper_bound) {
  std::uniform_int_distribution<int> RandomUniform(1, upper_bound);
  return RandomUniform(MersenneRandom);
}
```

gribozavr replied:
Right!

tinysun212 replied:
Could you choose a name instead of ```*_arc4random_*```? It will not be related to RC4 nor ```arc4random()``` no more. If the name is changed, other source ```./stdlib/private/SwiftPrivate/PRNG.swift``` will be changed together.

gribozavr replied:
How about `*_cxx11_mt19937_*`?

tinysun212 replied:
Thanks.

gribozavr commented on stdlib/public/runtime/ProtocolConformance.cpp:
```diff
@@ -565,7 +593,11 @@ swift::swift_conformsToProtocol(const Metadata *type,
+#if defined(_MSC_VER)
+  C.SectionsToScanLock.lock();
+#else
   pthread_mutex_lock(&C.SectionsToScanLock);
```
I don't feel good about the proliferation of `#if`s in concurrency code.  The code is complex as is.  Could you continue the work from https://github.com/apple/swift/pull/1731 and move the runtime over to platform-independent mutex wrappers?

tinysun212 replied:
I'll change this PR if #1731 is finally merged.

shawnce replied:
Note also the work-in-progress in PR https://github.com/shawnce/swift/pull/7. I only need to finish off some header docs and likely some spit and polish before opening a PR back into Apple's repo. I had hope #1731 would have been in by now.

gribozavr commented on stdlib/public/runtime/Metadata.cpp:
```diff
@@ -78,12 +96,20 @@ void *MetadataAllocator::alloc(size_t size) {
                       != (((uintptr_t)end & ~pagesizeMask)))){
+#if defined(_MSC_VER)
+    next = (char *)
+      VirtualAlloc(nullptr, pagesizeMask+1,
```
Could you define a wrapper over mmap()/VirtualAlloc() instead of duplicating the logic?

gribozavr commented on stdlib/public/runtime/Metadata.cpp:
```diff
@@ -59,17 +65,29 @@ using namespace metadataimpl;
 #if defined(__APPLE__)
   const uintptr_t pagesizeMask = vm_page_mask;
+#elif defined(_MSC_VER)
+  SYSTEM_INFO  SystemInfo;
```
Two spaces between type and variable name.

gribozavr commented on stdlib/public/runtime/CMakeLists.txt:
```diff
@@ -22,6 +22,10 @@ if("${CMAKE_SYSTEM_NAME}" STREQUAL "CYGWIN")
   set(swift_runtime_port_sources
       CygwinPort.cpp)
 endif()
+if("${CMAKE_SYSTEM_NAME}" STREQUAL "Windows")
```
Would it make sense to do:

```
if(("${CMAKE_SYSTEM_NAME}" STREQUAL "Windows") OR
   ("${CMAKE_SYSTEM_NAME}" STREQUAL "CYGWIN"))
```

instead of duplicating the logic?

tinysun212 commented:
Testing OS X Platform is failed. Is this my fault?

gribozavr commented:
@tinysun212 Currently master is passing tests on OS X, so it might be related.  Let's retest to be sure:

@swift-ci Please test

gribozavr commented on stdlib/public/runtime/Metadata.cpp:
```diff
@@ -27,9 +27,15 @@
 #include <new>
 #include <cctype>
+#if defined(_MSC_VER)
+#include <windows.h>
```
`#define NOMINMAX` please.

tinysun212 commented on stdlib/public/runtime/Casting.cpp:
```diff
@@ -32,6 +32,11 @@
+#else
+#include <pthread.h>
+#endif
 #include <type_traits>
```
I think I should expand ```Mutex.h/Mutex.cpp``` and use it here.

shawnce replied:
Please review the newly opened PR #1950. It adds read/write lock support as well is a little better factored to allow for alternate platforms to provide their own implementations if they don't have pthreads.

tinysun212 commented:
(mistake in rebasing.. I dropped this branch and restored.)
I added ```MutexWin32.cpp/h``` and revoked the ```Casting.cpp``` which is using read/write lock. Later, ```Mutex.h``` should be patched to include ```MutexWin32.h```.

gribozavr commented:
How does this work?  I'm not seeing any code using `ConditionPlatformHelper`.

I thought that you would modify `include/swift/Runtime/Mutex.h` and add `#if`s for Windows in `class Mutex` etc.

shawnce commented:
He is depending on my open PR #1950 which has the pthread equivalent
platform helper code.

On Sat, Apr 2, 2016 at 3:06 AM Dmitri Gribenko <notifications@github.com>
wrote:

> How does this work? I'm not seeing any code using ConditionPlatformHelper.
>
> I thought that you would modify include/swift/Runtime/Mutex.h and add #ifs
> for Windows in class Mutex etc.
>
> —
> You are receiving this because you commented.
>
> Reply to this email directly or view it on GitHub
> <https://github.com/apple/swift/pull/1918#issuecomment-204685686>
>

tinysun212 commented:
Yes, this PR should be merged with one more commit (Mutex.h patching) after #1950 merged. I'm waiting it is merged or other reviews.

shawnce commented on stdlib/public/runtime/MutexWin32.cpp:
```diff
@@ -0,0 +1,90 @@
+
+void ConditionPlatformHelper::wait(CONDITION_VARIABLE &condition,
+								   SRWLOCK &mutex) {
+  SleepConditionVariableSRW(&condition, &mutex, INFINITE, 0);
```
Looks like this is one of the few functions that return an error...

> If the function succeeds, the return value is nonzero.
> If the function fails, the return value is zero. To get extended error information, call GetLastError.

...consider testing for an error and calling fatalError like is done in MutexPThread.cpp.

shawnce commented on include/swift/Runtime/MutexWin32.h:
```diff
@@ -0,0 +1,74 @@
+
+  // The unsafe versions don't do error checking.
+  static void unsafeLock(MutexHandle &mutex) {
+	AcquireSRWLockExclusive(&mutex);
```
Over indented.

shawnce commented on stdlib/public/runtime/ProtocolConformance.cpp:
```diff
@@ -390,17 +394,25 @@ void swift::_swift_initializeCallbacksToInspectDylib(
+#if defined(_MSC_VER)
+  HMODULE handle;
 
+  if (!info->dlpi_name || info->dlpi_name[0] == '\0') {
```
I would drop the { } for the "if block" or add { } for else the "else block".

shawnce commented on stdlib/public/stubs/Assert.cpp:
```diff
@@ -33,7 +63,8 @@ _swift_stdlib_reportFatalErrorInFile(const char *prefix, intptr_t prefixLength,
   char *log;
-  asprintf(&log, "%.*s: %.*s%sfile %.*s, line %zu\n", (int)prefixLength, prefix,
+  swift_asprintf(&log, "%.*s: %.*s%sfile %.*s, line %zu\n",
+           (int)prefixLength, prefix,
```
Align parameters. The addition of "swift_" shifted things over.

Look into using git-clang-format on your changes. I am not sure of a quick guide of how to get it up and running, worked it out on my own but didn't take notes.

shawnce commented on stdlib/public/runtime/MutexWin32.cpp:
```diff
@@ -0,0 +1,90 @@
+//===--- Mutex.cpp - ------------------------------------------------------===//
```
It may make sense to hoist all of this implementation into the MutexWin32.h since it is so simple (leaving behind `ConditionPlatformHelper::wait` assuming you add fatalReporting to it).

kjpgit commented on stdlib/public/stubs/Stubs.cpp:
```diff
@@ -232,8 +240,19 @@ extern "C" uint64_t swift_float80ToString(char *Buffer, size_t BufferLength,
+  }
+  std::string Line;
+  std::getline(std::cin, Line);
+  *LinePtr = static_cast<char *>(std::malloc(Line.size() + 1));
```
missing a memcpy?

kjpgit commented on stdlib/public/stubs/Stubs.cpp:
```diff
@@ -232,8 +240,19 @@ extern "C" uint64_t swift_float80ToString(char *Buffer, size_t BufferLength,
+  return Line.size();
+#else
   size_t Capacity = 0;
   return getline(LinePtr, &Capacity, stdin);
```
I realize it's not this patch's fault, but passing LinePtr (which is an invalidated out param) directly into getline is not defensive coding.   Either use a temp var or assert LinePtr is null.

kjpgit commented on stdlib/public/stubs/Stubs.cpp:
```diff
@@ -232,8 +240,19 @@ extern "C" uint64_t swift_float80ToString(char *Buffer, size_t BufferLength,
+    return -1;
+  }
+  std::string Line;
+  std::getline(std::cin, Line);
```
IO error checking?

tinysun212 commented:
I applied reviews and rebased.

tinysun212 commented:
@shawnce, @kjpgit Could you review the fixed ```include/swift/Runtime/MutexWin32.h```, ```stdlib/public/stubs/Stubs.cpp```, and others?

karlito-go commented on stdlib/public/stubs/Stubs.cpp:
```diff
@@ -252,8 +258,43 @@ extern "C" uint64_t swift_float80ToString(char *Buffer, size_t BufferLength,
 /// if an error occurred, or EOF was reached.
 SWIFT_RUNTIME_STDLIB_INTERFACE
 extern "C" ssize_t swift_stdlib_readLine_stdin(char **LinePtr) {
+#if defined(_MSC_VER)
```
I still think that the first thing this function should do, for all platforms, is: *LinePtr = NULL.
It's an out parameter, according to the docstring, not in/out.

kjpgit replied:
I still think that the first thing this function should do, for all platforms, is: *LinePtr = NULL.  The docstring says thats an out param, not in/out.

tinysun212 replied:
I simply emulated the behavior of ```getline()```. Do you think we can ignore the input value of *LinePtr?

kjpgit commented on stdlib/public/stubs/Stubs.cpp:
```diff
@@ -252,8 +258,43 @@ extern "C" uint64_t swift_float80ToString(char *Buffer, size_t BufferLength,
+    if (Capacity - Pos <= 1) {
+      // Capacity changes to 128, 128*2, 128*4, 128*8, ...
+      Capacity = Capacity ? Capacity * 2 : 128;
+      ReadBuf = static_cast<char *>(realloc(ReadBuf, Capacity));
```
this leaks memory if realloc fails after first iteration.   not sure why you changed this func away from using std::getline?

tinysun212 replied:
If ```realloc()``` is failed, function returns -1 and caller should call ```free()``` with the parameter passed this function. The manual of ```getline()``` says that ```This buffer
       should be freed by the user program even if getline() failed.```.
Unfortunately, ```std::getline``` did not appened the end of line character, ```readLine(strippingNewline: false)``` did not work in swift.

tinysun212 replied:
I realized that ```readLine()``` in ```stdlib/public/core/InputStream.swift``` did not call ```_swift_stdlib_free()``` when ```swift_stdlib_readLine_stdin()``` is failed.

I think if we consider the fail of ```realloc()``` or ```getline()```, ```readLine()``` should be fixed or rewrite exsiting ```swift_stdlib_readLine_stdin()```.

kjpgit commented on stdlib/public/stubs/Stubs.cpp:
```diff
@@ -252,8 +258,43 @@ extern "C" uint64_t swift_float80ToString(char *Buffer, size_t BufferLength,
+      ReadBuf = static_cast<char *>(realloc(ReadBuf, Capacity));
+      if (ReadBuf == nullptr)
+        return -1;
+      *LinePtr = ReadBuf;
```
I think *LinePtr only needs to be set at the end of the func on success.

tinysun212 replied:
I keep the latest allocated pointer to give it to the caller even if this function is failed.

tinysun212 commented:
I have a plan create another PR related to https://bugs.swift.org/browse/SR-1280. It will modify ```swift_stdlib_readLine_stdin(char **LinePtr)``` and not use ```getline()```.

tinysun212 commented:
@kjpgit, some changes to ```swift_stdlib_readLine_stdin()``` modified as you pointed out.

tinysun212 commented:
Added ```MutexWin32.h``` to ```Mutex.h``` and patched ```CMaktLists.txt```.  @gribozavr, ready to review.

tinysun212 commented:
rebased and resolved conflicts

tinysun212 commented:
@gribozavr, rebased and resolved conflicts, this is ready to be reviewed, again.

gribozavr commented:
@swift-ci Please test

gribozavr commented:
This is a large patch, so I'll be leaving per-file LGTM comments, if you don't mind.

gribozavr commented on stdlib/public/SwiftShims/LibcShims.h:
```diff
@@ -31,6 +31,8 @@ namespace swift { extern "C" {
 #if defined(__linux__) && defined (__arm__) && !defined(__android__)
 typedef      int __swift_ssize_t;
+#elif defined(_MSC_VER)
+typedef long long __swift_ssize_t;
```
Changes to this file LGTM.

gribozavr commented on stdlib/public/runtime/CMakeLists.txt:
```diff
@@ -60,6 +67,8 @@ set(swift_runtime_sources
     Remangle.cpp
     swift_sections.S
+    MutexPThread.cpp
+    MutexWin32.cpp
```
Changes to this file LGTM.

gribozavr commented on stdlib/public/runtime/Private.h:
```diff
@@ -183,6 +183,8 @@ namespace swift {
                                    unsigned long *sectionSize);
+#endif
+#if defined(__CYGWIN__)
   void _swift_once_f(uintptr_t *predicate, void *context,
```
So the ABI of `std::once_flag` on MSVC happens to match what we expect?  If so, please extend the comment in `stdlib/public/runtime/Once.cpp` to mention that MSVC port also relies on this, not just libstdc++-based ports.

gribozavr replied:
Changes to this file LGTM otherwise.

tinysun212 replied:
```sizeof(swift_once_t) <= sizeof(void*)``` was passed. But I couldn't conform that  ```std::call_once()``` follows compatible init process (0 to 1 to 2). Could you let me know the counter part source which determines the compatibility.

gribozavr replied:
It is the code after `if (Builtin.ID == BuiltinValueKind::Once)` in `lib/IRGen/GenBuiltin.cpp`.

jckarter replied:
The compiler assumes that `swift_once_t` is zero-initialized, and nonzero after initialization finishes. It doesn't care what exact nonzero values it becomes once initialization begins.

tinysun212 replied:
The definition of ```std::once_flag``` on MSVC is
```
struct once_flag
	{	// opaque data structure for call_once()
	constexpr once_flag() noexcept
		: _Opaque(0)
		{	// default construct
		}
     .....
	void *_Opaque;
	};
```
.
I verified that the variable of the type ```std::once_flag``` is initialized to zero and the variable is changed to 2 after I call ```std::call_once``` with it. Does this behavior match what we expect?

gribozavr replied:
Yes, this matches our expectations.

gribozavr commented on stdlib/public/stubs/Assert.cpp:
```diff
@@ -89,9 +120,10 @@ _swift_stdlib_reportUnimplementedInitializer(const char *className,
+  swift_asprintf(&log, "fatal error: use of unimplemented "
+                       "initializer '%.*s' for class '%.*s'\n",
+                 (int)initNameLength, initName, (int)classNameLength,
+                 className);
```
Changes to this file LGTM.

gribozavr commented on stdlib/public/runtime/SwiftObject.mm:
```diff
@@ -34,7 +34,9 @@
 #include "../SwiftShims/RuntimeShims.h"
 #include "Private.h"
 #include "swift/Runtime/Debug.h"
+#if !defined(_MSC_VER)
```
The uses of `dlsym` are guarded by `#if SWIFT_OBJC_INTEROP`, so let's use the same condition here as well.

tinysun212 replied:
This will be fixed in next commit.

gribozavr commented on stdlib/public/runtime/Reflection.mm:
```diff
@@ -24,7 +24,10 @@
+#include <tuple>
+#if !defined(_MSC_VER)
 #include <dlfcn.h>
+#endif
```
This include does not seem to be used on all platforms, could you try removing it?

tinysun212 replied:
This will be fixed in next commit.

gribozavr commented on stdlib/public/stubs/UnicodeNormalization.cpp:
```diff
@@ -327,7 +327,7 @@ swift::_swift_stdlib_unicode_strToLower(uint16_t *Destination,
                                         int32_t SourceLength) {
   UErrorCode ErrorCode = U_ZERO_ERROR;
-#if defined(__CYGWIN__)
+#if defined(__CYGWIN__) || defined(_MSC_VER)
```
Changes to this file LGTM.

gribozavr commented on include/swift/Runtime/MutexWin32.h:
```diff
@@ -0,0 +1,95 @@
+};
+}
+
+#endif
```
This file LGTM.

jckarter replied:
Missing newline at end of file here.

tinysun212 replied:
This will be fixed in next commit.

gribozavr commented on stdlib/public/runtime/MutexWin32.cpp:
```diff
@@ -0,0 +1,32 @@
+               "'SleepConditionVariableSRW()' failed with error code %d\n",
+               errorcode);
+  }
+}
```
This file LGTM.

gribozavr commented on include/swift/Runtime/Mutex.h:
```diff
@@ -22,6 +22,8 @@
+#include "swift/Runtime/MutexWin32.h"
 #else
 #error "Implement equivalent of MutexPThread.h/cpp for your platform."
 #endif
```
Changes to this file LGTM.

gribozavr commented on stdlib/public/runtime/CygwinPort.cpp:
```diff
@@ -148,3 +148,4 @@ void swift::_swift_once_f(uintptr_t *predicate, void *context,
   } else
     swiftOnceMutex.unlock();
 }
+#endif
```
Changes to this file LGTM.

gribozavr commented on stdlib/public/runtime/Errors.cpp:
```diff
@@ -223,7 +234,13 @@ swift::fatalError(uint32_t flags, const char *format, ...)
+  vsprintf(log, format, args);
+#else
   vasprintf(&log, format, args);
+#endif
```
Could you factor this part into a helper like you did with `swift_asprintf`?

gribozavr commented on stdlib/public/runtime/Errors.cpp:
```diff
@@ -18,18 +18,24 @@
 #include <unistd.h>
+#endif
+#if !defined(_MSC_VER)
 #include <pthread.h>
```
I don't think we are using pthread for anything in this file, could you try removing the include?

tinysun212 replied:
This will be fixed in next commit.

gribozavr commented on stdlib/public/runtime/Errors.cpp:
```diff
@@ -188,11 +194,16 @@ reportOnCrash(uint32_t flags, const char *message)
   asl_log(NULL, NULL, ASL_LEVEL_ERR, "%s", message);
 #endif
-#if !defined(__CYGWIN__) && !defined(__ANDROID__)
+#if !defined(__CYGWIN__) && !defined(__ANDROID__) && !defined(_MSC_VER)
```
This condition (`!defined(__CYGWIN__) && !defined(__ANDROID__) && !defined(_MSC_VER)`) is becoming more complex and it is repeated in a few places in this file.  Could you factor it into a macro?  Something like

```c
#if !defined(__CYGWIN__) && !defined(__ANDROID__) && !defined(_MSC_VER)
#  define SWIFT_SUPPORTS_BACKTRACE_REPORTING 0
#else
#  define SWIFT_SUPPORTS_BACKTRACE_REPORTING 1
#endif
```

tinysun212 replied:
The definition for 0 and 1 shouldn't be swap ? Or using positive condition,
```
#if defined(__CYGWIN__) || defined(__ANDROID__) || defined(_MSC_VER)
#  define SWIFT_SUPPORTS_BACKTRACE_REPORTING 0
#else
#  define SWIFT_SUPPORTS_BACKTRACE_REPORTING 1
#endif
```

gribozavr replied:
You're right.

gribozavr commented on stdlib/public/runtime/HeapObject.cpp:
```diff
@@ -28,7 +28,9 @@
 #include <cstdio>
 #include <cstdlib>
+#if !defined(_MSC_VER)
 #include <unistd.h>
```
It seems like we are not using this include for anything, could you try removing it?

tinysun212 replied:
This will be fixed in next commit.

tinysun212 replied:
I confused the ```HeapObject.h``` with ```HeapObject.cpp```. #2829 should contain ```HeapObject.h``` instead of ```HeapObject.cpp```.

gribozavr commented on stdlib/public/runtime/Metadata.cpp:
```diff
@@ -56,19 +62,45 @@
+
+// free memory allocated by swift_allocPage()
+// On success, swift_freePage() returns 0, on failure -1
+static int swift_freePage(void *addr, size_t size) {
```
Seems like we are not checking the return value of this function.  Could you change it to return void and trap if the system call failed?

gribozavr commented on stdlib/public/runtime/Metadata.cpp:
```diff
@@ -56,19 +62,45 @@
+#elif defined(_MSC_VER)
+  SYSTEM_INFO SystemInfo;
+  GetSystemInfo(&SystemInfo);
+  const uintptr_t PageSizeMask = SystemInfo.dwPageSize - 1;
```
Could you factor the page size computation into a helper function, and cache the result there in a static variable?  Then it won't be recomputed every time around.

jckarter replied:
Rather than static initializations, we prefer the `SWIFT_LAZY_CONSTANT` macro, which uses more efficient `dispatch_once`-based lazy initialization.

gribozavr commented on stdlib/public/runtime/Metadata.cpp:
```diff
@@ -56,19 +62,45 @@
 using namespace metadataimpl;
 
+// allocate memory up to a nearby page boundary
+static void *swift_allocPage(size_t size) {
```
Could you rename this function pair to `swift_allocateMetadataRoundingToPage`/`swift_freeMetadata`?  Using "page" in the name makes it hard to interpret the parameter (is it size in bytes?  is it the number of pages?)  Also, we are passing the metadata tag on OS X, so this function should not be used for anything but metadata allocation.

I also see callers rounding the allocation size to a page size.  If this is critical for the function to operate correctly, could you move that rounding into the function?

jckarter replied:
The current implementation does assume that `swift_allocPage` returns page-aligned memory.

gribozavr commented on stdlib/public/runtime/MetadataLookup.cpp:
```diff
@@ -34,7 +34,11 @@
+#if defined(_MSC_VER)
+#include <windows.h>
+#else
 #include <dlfcn.h>
```
It seems like this header is only used on Apple systems.  Could you check?  If so, could you move it into the `#if defined(__APPLE__) && defined(__MACH__)` block above?

gribozavr commented on stdlib/public/runtime/MetadataLookup.cpp:
```diff
@@ -34,7 +34,11 @@
 #endif
 
+#if defined(_MSC_VER)
+#include <windows.h>
```
Just curious -- what do we need windows.h for in this file?

gribozavr commented on stdlib/public/runtime/ProtocolConformance.cpp:
```diff
@@ -396,17 +400,25 @@ void swift::_swift_initializeCallbacksToInspectDylib(
+  if (!info->dlpi_name || info->dlpi_name[0] == '\0')
+    handle = GetModuleHandle(nullptr);
+  else
+    handle = GetModuleHandle(info->dlpi_name);
```
Can Cygwin use `GetModuleHandle` too?

tinysun212 replied:
Yes, Cygwin can use both functions. I didn't merge because Cygwin prefers ```POSIX``` version.

gribozavr commented:
```c
__swift_ssize_t
 swift::_swift_stdlib_write(int fd, const void *buf, __swift_size_t nbyte) {
   return write(fd, buf, nbyte);
 }
```

Does this work?  In another place you had to change `write()` to `_write()`.  What's the difference between the two?

gribozavr commented:
@swift-ci Please smoke test OS X platform

gribozavr commented on include/swift/Runtime/HeapObject.h:
```diff
@@ -106,8 +106,8 @@ struct TwoWordPair {
-#if __arm__ || __i386__ || defined(__CYGWIN__)
-#if defined(__CYGWIN__)
+#if __arm__ || __i386__ || defined(__CYGWIN__) || defined(_MSC_VER)
+#if defined(__CYGWIN__) || defined(_MSC_VER)
```
Changes to this file LGTM.

gribozavr commented on stdlib/public/stubs/LibcShims.cpp:
```diff
@@ -70,6 +79,11 @@ size_t swift::_swift_stdlib_malloc_size(const void *ptr) {
+#include <malloc.h>
+size_t _swift_stdlib_malloc_size(const void *ptr) {
+  return _msize(const_cast<void *>(ptr));
+}
```
Changes to this file LGTM.

gribozavr replied:
Changes to this file LGTM.

gribozavr commented on stdlib/public/stubs/LibcShims.cpp:
```diff
@@ -12,11 +12,16 @@
 #include <stdio.h>
 #include <string.h>
 #include "../SwiftShims/LibcShims.h"
+#include "llvm/Support/DataTypes.h"
```
What is this for?

tinysun212 replied:
MSVC doesn't have ```ssize_t``` which is not part of C++11.  I used ```DataType.h``` for definition of ```ssize_t```.

gribozavr replied:
I see.  But... what is the return type of `read()` and `write()`?

tinysun212 replied:
```_read()``` and ```_write()``` returns ```int```.

gribozavr commented on stdlib/public/stubs/Stubs.cpp:
```diff
@@ -249,13 +255,60 @@ extern "C" uint64_t swift_float80ToString(char *Buffer, size_t BufferLength,
+  if (LinePtr == nullptr)
+    return -1;
+
+  int Capacity = 0;
```
Probably does not matter in practice, but `Capacity` should probably be declared as `ssize_t`.

gribozavr commented on stdlib/public/stubs/Stubs.cpp:
```diff
@@ -249,13 +255,60 @@ extern "C" uint64_t swift_float80ToString(char *Buffer, size_t BufferLength,
+      break;
+  }
+
+  ReadBuf[Pos] = '\0';
```
Wouldn't this overflow the buffer in case `Capacity - Pos` was 2 in the check above?

tinysun212 replied:
There are three control path after the condition of ```Capacity - Pos >= 2```.
1) ```ch == EOF``` and exit loop. The buffer will be used for ```'\0'``` terminating. (1 Byte)
2) ```ch == '\n'``` and exit loop. The buffer will be used for non-EOF character and  ```'\0'``` terminating. (2 Bytes)
3) ```ch != EOF``` and ```ch != '\n'``` and continue loop. The buffer will be used for non-EOF character. (1 Byte)

I believe the buffer will not be overflowed.

gribozavr replied:
I see now, thanks for the explanation.

gribozavr commented on stdlib/public/stubs/Stubs.cpp:
```diff
@@ -453,17 +506,31 @@ const char *swift::_swift_stdlib_strtof_clocale(
+  return SystemInfo.dwNumberOfProcessors;
+#else
   return sysconf(_SC_NPROCESSORS_ONLN);
+#endif
```
Changes to this file LGTM otherwise.

gribozavr replied:
Changes to this file LGTM.

gribozavr commented:
@tinysun212 Please consider submitting approved parts as a separate pull request.  This will make it much easier to iterate on the other parts.

tinysun212 commented:
@gribozavr Thanks for your reviews, I'll make separate PR containing only approved parts.

jckarter commented on stdlib/public/runtime/CMakeLists.txt:
```diff
@@ -18,9 +18,17 @@ if(SWIFT_RUNTIME_ENABLE_LEAK_CHECKER)
+if("${CMAKE_SYSTEM_NAME}" STREQUAL "Windows")
   set(swift_runtime_port_sources
+      MutexWin32.cpp
       CygwinPort.cpp)
```
Is CygwinPort.cpp intended to be part of the MSVC build? Should it be renamed `Win32.cpp` or something more general instead?

tinysun212 replied:
Yes, it also used for the MSVC build. I agree it to have a new name, but leaved for another PR or iteration.

gribozavr commented:
@tinysun212 Thank you, I merged #2829.  Would be happy to continue to iterate on the rest!

tinysun212 commented:
@gribozavr, I rebased without any patch for the convenience. Your reviews will be applied in several commits.

gribozavr commented on stdlib/public/runtime/Reflection.mm:
```diff
@@ -24,6 +24,9 @@
 #include <string>
+#if !defined(_MSC_VER)
+#include <dlfcn.h>
+#endif
```
You shouldn't need this :)

gribozavr commented on stdlib/public/runtime/MetadataLookup.cpp:
```diff
@@ -34,6 +34,12 @@
+#if defined(_MSC_VER)
+#include <windows.h>
+#else
+#include <dlfcn.h>
```
You shouldn't need `dlfcn.h` here :)

tinysun212 committed:
Modified for some reviews

tinysun212 commented:
@gribozavr, @jckarter, I'll touch all your previous reviews one by one, but it take some time. You can know if it is touched from my comment for your review.

gribozavr commented:
@tinysun212 Just let us know when you're ready!

jckarter commented:
@gribozavr `_read` and `_write` are more-or-less equivalent to the POSIX APIs. Microsoft consistently marks non-standard APIs in their C runtime with a leading underscore.

tinysun212 commented:
In ```stdlib/public/stubs/LibcShims.cpp```, ```_read/_write/_close``` should be used for MSVC. The patch was omitted for some reason.

tinysun212 commented:
@gribozavr, this PR is ready for review.

gribozavr commented on stdlib/public/runtime/Errors.cpp:
```diff
@@ -50,7 +60,7 @@ enum: uint32_t {
 using namespace swift;
 
-#if !defined(__CYGWIN__) && !defined(__ANDROID__)
+#if SWIFT_SUPPORTS_BACKTRACE_REPORTING
```
Thank you, this is much cleaner now!

gribozavr commented on stdlib/public/runtime/Once.cpp:
```diff
@@ -49,6 +49,8 @@ void swift::swift_once(swift_once_t *predicate, void (*fn)(void *)) {
+  // follows the compatible init process.
   // For more information, see rdar://problem/18499385
   std::call_once(*predicate, [fn]() { fn(nullptr); });
 #endif
```
Changes to this file LGTM.

gribozavr commented on stdlib/public/runtime/Private.h:
```diff
@@ -183,6 +183,8 @@ namespace swift {
+#if defined(__CYGWIN__)
   void _swift_once_f(uintptr_t *predicate, void *context,
                      void (*function)(void *));
 #endif
```
Changes to this file LGTM.

gribozavr commented on stdlib/public/runtime/ProtocolConformance.cpp:
```diff
@@ -416,7 +428,11 @@ static int _addImageProtocolConformances(struct dl_phdr_info *info,
+#endif
   return 0;
 }
 
```
Changes to this file LGTM.

gribozavr commented on stdlib/public/runtime/SwiftObject.mm:
```diff
@@ -34,7 +34,9 @@
+#endif
 #include <stdio.h>
 #include <stdlib.h>
 #include <unordered_map>
```
Changes to this file LGTM.

gribozavr commented:
All changes LGTM.

gribozavr commented:
@swift-ci Please test and merge
